feat(ast)!: add `value` field to `BigIntLiteral`

overlookmotel opened the pull request:
Add `value` field to `BigIntLiteral`, same as the other literal types.

Once upon a time, we stored the value as a `BigInt` which normalized the representation (removing `_`s, and converting e.g. `0xFF` to `255`). But we had to remove it because `BigInt` owns a heap allocation, and this caused a memory leak.

https://github.com/oxc-project/oxc/commit/50fe0fef73027f10740e68a65a78216933b5d8a0

This PR brings back the `value` field, but avoids memory leak by storing the normalized value as a string (`Atom`).

This has effects in various places in the codebase:

* Brings our JS-side AST into line with ESTree (fixing the test cases broken by #11563).
* Fixes edge cases in a couple of lint rules.
* Codegen now prints bigints in normalized form e.g. `0xFFn` is printed as `255n`.
* Changes behavior of 1 function in `oxc_ecmascript`.

I've made comments below on the parts I'm not so sure about. Review would be very welcome!

overlookmotel commented on crates/oxc_semantic/src/dot.rs:
```diff
@@ -45,7 +45,7 @@ impl DebugDotContext<'_, '_> {
             AstKind::BooleanLiteral(lit) => Some(lit.value.to_string()),
             AstKind::StringLiteral(lit) => Some(lit.value.to_string()),
-            AstKind::BigIntLiteral(lit) => Some(lit.raw.to_string()),
+            AstKind::BigIntLiteral(lit) => Some(lit.value.to_string()),
```
Anyone have any idea if this change is correct? For bigint `0x1_0n`, previously would have returned `"0x1_0n"`, now it returns `"16"`.

But I have no idea what this `DebugDotContext` thing is! (I assume part of CFG, but that's a part of the codebase I know nothing about)

camc314 replied:
i think this is fine.

this function is just a debug representation of the CFG and isn't used anywhere

overlookmotel commented on crates/oxc_ecmascript/src/to_string.rs:
```diff
@@ -97,9 +96,7 @@ impl<'a> ToJsString<'a> for NumericLiteral<'a> {
-        self.base
-            .is_base_10()
-            .then(|| Cow::Owned(self.raw.trim_end_matches('n').cow_replace('_', "").to_string()))
+        Some(Cow::Borrowed(self.value.as_str()))
```
I assume the original base is not relevant here, and testing for it was just a quick workaround?

overlookmotel commented on crates/oxc_codegen/src/gen.rs:
```diff
@@ -1284,15 +1284,14 @@ impl GenExpr for BigIntLiteral<'_> {
             p.print_ascii_byte(b'(');
-            p.print_str(&raw);
-            p.print_ascii_byte(b')');
+            p.print_str(value);
+            p.print_str("n)");
```
Are you on crack, Mr bot?

overlookmotel committed:
feat(ast)!: add `value` field to `BigIntLiteral` (#11564)

Add `value` field to `BigIntLiteral`, same as the other literal types.

Once upon a time, we stored the value as a `BigInt` which normalized the representation (removing `_`s, and converting e.g. `0xFF` to `255`). But we had to remove it because `BigInt` owns a heap allocation, and this caused a memory leak.

https://github.com/oxc-project/oxc/commit/50fe0fef73027f10740e68a65a78216933b5d8a0

This PR brings back the `value` field, but avoids memory leak by storing the normalized value as a string (`Atom`).

This has effects in various places in the codebase:

* Brings our JS-side AST into line with ESTree (fixing the test cases broken by #11563).
* Fixes edge cases in a couple of lint rules.
* Codegen now prints bigints in normalized form e.g. `0xFFn` is printed as `255n`.
* Changes behavior of 1 function in `oxc_ecmascript`.

I've made comments below on the parts I'm not so sure about. Review would be very welcome!
